File: src/algorithms/arrandstring/GasStation.java
package algorithms.arrandstring; 

import java.util.*; 

public class GasStation {
    /* Function: Given an array shows the gas each station contains, another array cost[i] shows 
     *     that from i station to i + 1 need how much gas. Find the start point that can complete
     *     the circuit. 
     * Solution: left+gas[i]-cost[i] means when it reach i station, how much gas it has. So if it 
     *     is nagetive, this station cannot be the starter. 
     */ 
    
    public int canCompleteCircuit(int[] gas, int[] cost) {
        if (gas == null || gas.length == 0 || cost == null || cost.length != gas.length) {
            return -1;
        }
        
        int val = 0, sum = 0; 
        int res = -1; 
        for (int i = 0; i < gas.length; ++i) {
            val += gas[i] - cost[i]; 
            sum += gas[i] - cost[i]; 
            if (sum < 0) {
                sum = 0; 
                res = i; 
            }
        }
        
        if (val < 0) {
            return -1; 
        }
        
        return res + 1; 
    }
}

File: src/CPP/src/leetcode/GasStation.cpp
/*
 * [Source] https://leetcode.com/problems/gas-station/
 * [Difficulty]: Medium
 * [Tag]: Greedy
 */

#include <iostream>

using namespace std;

// [Solution]: If pre + gas[i] < cost[i], then set i+1 as starter. Until reach starter again. Since if we cannot reach one station, it means starter cannot be any previous station, as we always have a greater-than-0 pre.
// [Corner Case]: Cannot reach starter again, return -1.
class Solution {
};

// [Solution]: If cannot finish the circle, the sum of gas[i] < sum of cost[i]. Use this to pass the array only once.
/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/arrandstring/GasStation.java
 */

/* Java solution
public class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
		if(gas.length!=cost.length) return -1;
		int start=-1; // The valid start point. 
		int leftGas=0; // If <0, then cannot start at current station. 
		int total=0;
		for(int i=0; i<gas.length; i++){
			total+=gas[i]-cost[i];
			leftGas+=gas[i]-cost[i];
			if(leftGas<0){
				start=i;
				leftGas=0;
			}
		}
		if(total<0) return -1;
		else return start+1;
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
